notebooks/ch12/LR_Case_Study.md

```diff
@@ -0,0 +1,792 @@
+
+
+```python
+# HIDDEN
+import numpy as np
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
+from sklearn.linear_model import LinearRegression
+%matplotlib inline
+
+sns.set()
+sns.set_context('talk')
+pd.options.display.max_rows = 7
+pd.options.display.max_columns = 7
+np.set_printoptions(precision=2, suppress=True)
+```
+
+
+```python
+# HIDDEN
+"""
+def minimize(cost_fn, grad_cost_fn, X, y, alpha=0.2, progress=True):
+    '''
+    Uses gradient descent to minimize cost_fn. Returns the minimizing value of
+    theta once theta changes less than 0.001 between iterations.
+    '''
+    theta = np.zeros(X.shape[1])
+    old_cost = 10000000
+    while True:
+        if progress:
+            print(f'theta: {theta} | cost: {cost_fn(theta, X, y):.2f}')
+        gradient = grad_cost_fn(theta, X, y)
+        new_theta = theta - alpha * gradient
+        cost = cost_fn(theta, X, y)
+        if abs(old_cost - cost) < 10:
+            return new_theta
+        old_cost = cost
+        
+        theta = new_theta
+        
+"""
+        
+from scipy.optimize import minimize as sci_min
+
+def minimize(cost_fn, grad_cost_fn, X, y, progress=True):
+    '''
+    Uses scipy.minimize to minimize cost_fn using a form of gradient descent.
+    '''
+    theta = np.zeros(X.shape[1])
+    iters = 0
+    
+    def objective(theta):
+        return cost_fn(theta, X, y)
+    def gradient(theta):
+        return grad_cost_fn(theta, X, y)
+    def print_theta(theta):
+        nonlocal iters
+        if progress and iters % progress == 0:
+            print(f'theta: {theta} | cost: {cost_fn(theta, X, y):.2f}')
+        iters += 1
+        
+    print_theta(theta)
+    return sci_min(
+        objective, theta, method='BFGS', jac=gradient, callback=print_theta,
+        tol=1e-7
+    ).x
+```
+
+## Linear Regression Case Study
+
+In this section, we will perform an end-to-end case study of the application of a linear regression model to a dataset.
+
+The dataset we will be working with is a record of the population and drinking data in various states. The dataset is from this website: http://people.sc.fsu.edu/~jburkardt/datasets/regression/x20.txt, with some properties of the dataset changed for didactical purposes.
+
+Our task in this section: fit a good linear regression model to this dataset in order to predict the cirrhosis death rate.
+
+## Exploring the Dataset
+
+As in any data science project, we're going to begin by doing some exploratory data analysis.
+
+We'll start by reading in the dataset using `pandas`.
+
+
+```python
+drinking_data = pd.read_csv("drinking.csv")
+```
+
+It's always a good idea to look at _how much_ data we have by looking at the number of columns and the number of rows in our dataset.
```

Comment:
number of columns and the number of rows in our -> dimensions of the

Reply:
also can you mention why it's a good idea? like if its 999999999 rows don't print the whole df?